### ISSUE TITLE
accessory.name displayName and uuid generation

### ISSUE THREAD

**Issue by justme-1968** [nfarina/homebridge#203]
in the old platform support issue #95 there was a short discussion about the generation of the uuid for each accessory. unfortunately we have not followed this to the end.

for devices that have a real unique serial number it would be better to create the uuid based on this serial number an not on the (display) name which might change during the lifetime of the device.

i would like to suggest creating the (platform) accessories with something like this:
```
var accessoryUUID = uuid.generate(accessoryInstance.constructor.name + ":" + (accessoryInstance.uuid_base || displayName));
```

instead of a new variable uuid_base we could also use the serialNumber from the AccessoryInformation service.

**Comment by snowdd1** [comment]
You mean here
https://github.com/nfarina/homebridge/blob/master/app.js#L162
?
I would support that. During modeling my house I had to change names quite often, as in a global (or house-wide) context the name was not distinctive enough, loosing room asignments and such.
I assume with an constant UUID that would not have happened.

Did you test your change in you own fork/locally?

AccessoryInformation service is quite global in my platform, I would rather configure them as a new property with the accessories within.

**Comment by justme-1968** [comment]
yes. in app.js.

i will check if it works as intended.

**Comment by desvdp** [comment]
see also #148

**Comment by justme-1968** [comment]
what exactly is accessoryInstance.constructor.name in the call to uuid.generate supposed to do?
as far as i can see it just prefixes everything with an 'Object' string. so it looks redundant.

so this:
```
var accessoryUUID = uuid.generate(accessoryInstance.uuid_base || displayName);
```
should be enough?

**Comment by desvdp** [comment]
Actually it prefixes with the name of the accessory type you're generating a UUID for, making accessories unique by their accessory shim by getting the name of the constructor object

> function MyCoolAccessory(n) { this.name=n;}
> var myThing = new MyCoolAccessory("home");
> console.log(myThing.constructor.name);

--> MyCoolAccessory

**Comment by justme-1968** [comment]
hmmm... i'm not sure if this works as intended. if i put a debug message into uuid.generate it just prefixes everything with 'Object'.

**Comment by snowdd1** [comment]
Are you using object literals?

**Comment by justme-1968** [comment]
what exactly do you mean ?

**Comment by nfarina** [comment]
If you're talking about the FHEM platform, then it looks like it should be saying things like `FHEMAccessory` instead of `Object`. Even if you were using object literals the result of calling `accessories()` are still instances of the `FHEMAccessory` class. So it's really weird that it's prefixing Object.

**Comment by justme-1968** [comment]
yes. i'm talking about FHEM.js.

**Comment by andig** [comment]
According to http://stackoverflow.com/questions/332422/how-do-i-get-the-name-of-an-objects-type-in-javascript, this: 

    var accessoryUUID = uuid.generate(accessoryInstance.constructor.name + ":" + (accessoryInstance.uuid_base || displayName));

will- as observed- always be prefixed by `Object` which is not intended (and not helpful).

> Actually it prefixes with the name of the accessory type you're generating a UUID for, making accessories unique by their accessory shim by getting the name of the constructor object

@desvdp Exactly that is what should be happening but doesn't. Your example

    function MyCoolAccessory(n) { this.name=n;}
    var myThing = new MyCoolAccessory("home");
    console.log(myThing.constructor.name);

Is not using the protoype pattern that the accessories use:

	function HttpAccessory(log, config) {
	}

	HttpAccessory.prototype = {
	}

I'm lost though how to fix this without forcing Accessories to store their type in another field?

**Comment by snowdd1** [comment]
I've just `console.log`ged my UUID creation and it is reliably `object` as constructor name.
I followed the trace back to the point where I create the accessory objects and even there the contructor name is already printed as `object`.

So @andig  is right:

````javascript
// testing constructor name

function MyCoolAccessory(n) { this.name=n;}
var myThing = new MyCoolAccessory("home");
console.log("1 " + myThing.constructor.name);

function My2CoolAccessory(n) { this.name=n;}
My2CoolAccessory.prototype = {
		a:0,
		b:1,
		c:function(){this.d = a+b;}
}
var my2Thing = new My2CoolAccessory("home");
console.log("2 " + my2Thing.constructor.name);
````
returns
````
$node test.js
1 MyCoolAccessory
2 Object
````
In the moment we are manipulating the prototype of the function, we lose the name, and it's just `Object`

**Question is, why do we do that?**
Being -still- a newbie to javascript I would think we could create our accessories that way (taking the example from above):
````javascript
function My3CoolAccessory(n) { 
		this.name=n;
		this.a = 0;
		this.b = 1;
		this.c = function (){this.d = this.a+this.b;};
};
var my3Thing = new My3CoolAccessory("home");
console.log("3:" + my3Thing.constructor.name);
````
returns `3:My3CoolAccessory` 

So instead of manipulating the prototype we could just define all methods we need within the constructor. Through prototypal inheritance  these methods are available in derived objects.

Am I completely wrong?

Another subject might be the use of `require()` which might also kill the prototype.name, but I am not sure of that yet.

**Comment by snowdd1** [comment]
With require() it gets even more complicated.
Taking the example from above, and exporting the My3CoolAccessory, we get in another file
````javascript
var a = require("./test.js").My3CoolAccessory;
console.log("a: " + a.constructor.name);

var b = new a("Hey");
console.log("b: " + b.constructor.name);
````
guess what?
Right:
````
a: Function
b: My3CoolAccessory
````

**Comment by nfarina** [comment]
@andig Thanks for the StackOverflow link, looks like `constructor.name` is not a good choice here. It worked for me but does fall down in many cases.

**Comment by andig** [comment]
> With require() it gets even more complicated.

Well why make it so complicated? E.g. in https://github.com/nfarina/homebridge/blob/master/app.js#L80 you already have the constructor name. Why not just pass it to `createAccessory` as `namespace` parameter?

**Comment by snowdd1** [comment]
Coming back to the original post of @justme-1968 :
What about having an `namespace` parameter for platforms/accessories and an optional `uuid_base` in the config file? This would allow changing the name without creating a new UUID and thus losing all room assignments and such.

If you agree (espc. @nfarina  ) I would prepare a PR.

**Comment by nfarina** [comment]
Yes I think that's a good idea, I would support that,

**Comment by snowdd1** [comment]
Ok, here it is.
Please have a look at it before merging #233 because I currently have no node compiler at hand to test.

**Comment by snowdd1** [comment]
Ok. it's tested with node 4.1.1 on my raspi.

**Comment by justme-1968** [comment]
i could not follow the diskussion during the last few days. but this sounds great.

**Comment by andig** [comment]
> I'm reopening this since according to nfarina/homebridge#203 this should actually work but doesn't.

After some checking I've found that homebridge does **NOT** handle creating accessories of same name if they have different type:

    accessories.push(new FritzOutletAccessory(self, "foo"));
    accessories.push(new FritzThermostatAccessory(self, "foo"));
    callback(accessories);

Gives:

	/Users/andig/Documents/htdocs/homebridge/node_modules/hap-nodejs/lib/Accessory.js:258
	      throw new Error("Cannot add a bridged Accessory with the same UUID as another bridged Accessory: " + existing.UUID);
	      ^

For `homebridge-fritz` this can be solved by adding the AIN as `uuid_base` parameter to `FritzPlatform.prototype` which I'll implement shortly.

Fixed in `0.5.8` and finally closing :)

**Comment by andig** [comment]
The reason is `Server.prototype._loadPlatformAccessories()`:

	var accessory = this._createAccessory(accessoryInstance, accessoryName, platformType, accessoryInstance.uuid_base);

and then in `_createAccessory()`:

    var accessoryUUID = uuid.generate(accessoryType + ":" + (uuid_base || displayName));

This only takes `accessoryType` and `displayName` into account unless `uuid_base` has been manually added to the accessory. `accessoryType` is afaiks the platform type and **NOT** the accesory type (at least with my implementation), so it creates duplicates.
